tlatools/org.lamport.tlatools/src/tlc2/debug/TLCBreakpointExpression.java

```diff
@@ -0,0 +1,269 @@
+		} while (definedNames.occurSymbol(unusedName));
+		return unusedName;
+	}
+	
+	/**
+	 * The expression operator can accept parameters; for example if we want
+	 * to add a breakpoint at the following location:
+	 *
+	 * VARIABLE i
+	 * op(j) ==
+	 *   \A k \in S :
+	 *     F(i, j, k) (* want to add breakpoint here)
+	 *
+	 * It would be nice for the user to just be able to type an expression
+	 * like "i + j >= k"; however, there then arises the problem of how to take
+	 * that expression and make it parse-able. Thus we end up auto-generating
+	 * an operator that looks like:
+	 *
+	 * __BreakpointExpr__0(j, k) == i + j >= k
+	 *
+	 * Where we put the two scoped identifiers as parameters to the operator
+	 * (note that we do not put the global identifier "i" as a parameter,
+	 * because SANY would raise a parse error for multiple "i" definitions).
+	 * Then, when evaluating the breakpoint expression we find those named
+	 * values in the stack frame and plug them in. It is tricky but possible
+	 * to isolate only the variables local to the breakpoint scope; that is
+	 * what this function does.
+	 *
+	 * However, this gives rise to another problem; what if the user writes a
+	 * spec like this?
+	 *
+	 * op(i) ==
+	 *   \A j \in S :
+	 *     F(i, j) (* want to add breakpoint here *)
+	 * VARIABLE i
+	 *
+	 * Since when we extend the module so as to define our generated op it is
+	 * implicitly parsed at the end of the file, "__BreakpointExpr__0(i, j)"
+	 * will result in a parse error due to multiple "i" definitions.
+	 * Currently these types of specs are simply not supported. Supporting
+	 * them would probably require changes to the semantic analysis and
+	 * constants-resolver code enabling parsing "ghost" definitions at
+	 * arbitrary points in the tree.
+	 *
+	 * The way this method works is it first finds the parse tree node where
+	 * the breakpoint was introduced, then traces its parentage up to the
+	 * module root while recording identifiers produced by:
+	 *  - LET/IN blocks
+	 *  - Quantification operators
+	 *  - Operator parameters
+	 *
+	 * There are probably others but we will add to this code as the need is
+	 * discovered. A better solution would be to record this info in the
+	 * semantic nodes during the semantic checking process but that will have
+	 * to wait until we have good test coverage of it. It should be noted
+	 * this information is already sort of recorded in the (private)
+	 * OpApplNode.allParams field, but only if the given expression actually
+	 * references an outside parameter. By the time the semantic analysis
+	 * process has actually reached that stage it's too late to append the
+	 * references to the operator definition node.
+	 *
+	 * @param semanticRoot The root node of the semantic parse tree.
+	 * @param location     The breakpoint location.
+	 * @return A list of local identifiers accessible at the given location.
+	 */
+	static Set<String> getScopedIdentifiers(
```

Comment:
Should/Could this be a method of `ModuleNode`?

Reply:
I think it's a bit half-baked to move into ModuleNode. I would rather fix this by refactoring each node to keep a snapshot of the `tla2sany.semantic.SymbolTable` stack around. Not something that can be done without the semantic test corpus.

Reply:
Out of scope for this PR.